### ISSUE TITLE
Empty .NET 7 app allocates huge amount of memory (512 MB / 4 GB) on Window 7 64 bit . 

### ISSUE THREAD

**Issue by ymalich** [dotnet/runtime#79469]
### Description

Hello community.
I've ported my application on .NET 7 and by testing it on Windows 7 64 bit I've descovered a bug - the .NET 7 process allocated huge amount on memory just in Main.
I've created a simple "Hello word" Console app and reproduced the problem on several test PC, so I sure it's a bug in .NET 7.0 runtime. 
Look 
1. In VS 2022 simple console app allocates 572 MB just in Main().
![00](https://user-images.githubusercontent.com/52053585/206781897-7277e629-e451-4bc9-a890-66ded7c3ef32.png)
![01](https://user-images.githubusercontent.com/52053585/206781947-d8ea35ba-98c2-41a7-a52e-7bf42e66fc1e.png)
2. Windows Task Manager shows the Commit GB grows by 4 GB after Console app starts! 
![03](https://user-images.githubusercontent.com/52053585/206780033-ca634642-33f6-42a2-9fc3-a6fd0de3fe29.png)
![02](https://user-images.githubusercontent.com/52053585/206780070-d79c367d-9cff-49eb-8b13-1765ddbec049.png)
3. Windows Task Manager shows Commit size 539 MB in the table 
![04](https://user-images.githubusercontent.com/52053585/206780230-5f61e043-650f-4641-8255-a62e79da0355.png)
4. Sysinternals Process Explorer - Virtual size is 270 983 436 KB!
![05](https://user-images.githubusercontent.com/52053585/206780292-ac32d586-f0b9-42a9-bfd5-1e86cb2e6391.png)
5. If the app runs under .NET6, it consumes only 14 MB!
6. ![07](https://user-images.githubusercontent.com/52053585/206782306-07163f28-8742-4a44-9374-9158eaaa6693.png)
[ConsoleAppWin7Test1.zip](https://github.com/dotnet/runtime/files/10197680/ConsoleAppWin7Test1.zip)

If it runs on Windows 7 32 bit the memory allocation is not much. 

Yes, I understand, the Windows 7 is not officially supported anymore (despite ESU) but a lot of people still use it and I suppose it can be a not Windows 7 specific bug but the hidden bug, that shows itself on Windows 7 64 bit for some internal reasons.
So I suppose it's better to find and fix it in all cases. 

### Reproduction Steps

Run any console app on Windows 7 64 bit.

### Expected behavior

Small memory consumption.

### Actual behavior

It allocates 4 GB memory.

### Regression?

Yes, it's not reproducible on .NET 6.0

**Comment by filipnavara** [comment]
FWIW I think you are comparing different things. I would expect the VS reported memory to be closer to the "Private Bytes" in Process Explorer. It appears to be in the similar ballpark, at very least.

**Comment by filipnavara** [comment]
> I'm just showing, what different tools reported about .NET 7.0 process memory allocation.

I'm just referring to the part you marked "If the app runs under .NET6, it consumes only 14 MB!" which references a totally different tool that reports a different thing. The closes equivalent in the screenshots would be the 7MB "Private Bytes" value in Process Explorer.

> I would say it's a bug that should be analyzed and fixed. :)

It should definitely be analyzed. I am not necessarily convinced it's a regression or bug but it could very well be.

**Comment by sebastienros** [comment]
@ymalich can you try again by setting the environment variable `COMPlus_GCName` to `clrgc.dll`? At least for this process.

**Comment by ymalich** [comment]
> It might be caused by the double allocator introduced by W^X. If you run your app on Linux you may see the virtual address space mapped is 2TB. Can you try setting the environment variable `DOTNET_EnableWriteXorExecute=0`?

Yes, it has helped!

**Comment by ymalich** [comment]
Hallo guys @sebastienros  and @hez2010 ,
neither setting COMPlus_GCName or DOTNET_EnableWriteXorExecute alone resolve the problem completely.
But setting both COMPlus_GCName and DOTNET_EnableWriteXorExecute together does it!

I'll send the whole report in several minutes.

**Comment by ymalich** [comment]
The report.
1. Memory usage in Task Manager before launching with set COMPlus_GCName=clrgc.dll
Memory = 3,05 GB, Commit (GB) =5, 
![10](https://user-images.githubusercontent.com/52053585/206842521-4e2ec3c9-bc3a-42c4-98e6-dd98b2b8fc43.png)

2. Memory usage in Task Manager after the app was launched with set COMPlus_GCName=clrgc.dll 
![11](https://user-images.githubusercontent.com/52053585/206843379-ea9bad18-d392-4336-a779-8aa555eb50f1.png)
Memory = 4,05 GB, Commit (GB) =9

4. But the process Commit size on the processes tab is 8 444 K
![12](https://user-images.githubusercontent.com/52053585/206842655-6ee970f4-e9b4-47c2-b01d-77a81ff262a6.png)

5. Process Explorer Private Bytes - 8 444 K, Virtual size 623 252 K (same picture for .NET 6.0 app). 
![13](https://user-images.githubusercontent.com/52053585/206843049-ba23ad16-17a9-4d3f-b723-b198f20eee57.png)
In Process Explorer it looks better than before.

Now I'll  DOTNET_EnableWriteXorExecute=0
6.  Memory usage in Task Manager before launching with set DOTNET_EnableWriteXorExecute=0
Memory = 3,20 GB, Commit (GB) =5, 
![14](https://user-images.githubusercontent.com/52053585/206842776-fefe0212-15d4-4d82-9583-fb7bc75cf1af.png)

7. Memory usage in Task Manager after the app was launched with set DOTNET_EnableWriteXorExecute=0
Memory = 3,21 GB, Commit (GB) =6. 
![15](https://user-images.githubusercontent.com/52053585/206843066-0b546768-4b8c-4e8c-ab1c-3cf39b1e644e.png)
It seems look good but.... not completely

8. But the process Commit size on the processes tab is 540 020 K !
![16](https://user-images.githubusercontent.com/52053585/206844835-4fbb52bd-4820-4610-b1b7-46a3ec2ea91e.png)

9. Process Explorer Private Bytes - 540 020 K, Virtual size 270 694 628 K
![17](https://user-images.githubusercontent.com/52053585/206845173-1b4ff083-11bb-4bf9-8db7-bc0ccf90e8b8.png)
It look here the same as before setting env variable.

10. Now I set both 
set COMPlus_GCName=clrgc.dll 
set DOTNET_EnableWriteXorExecute=0
Memory usage in Task Manager 
Memory = 3,47 GB, Commit (GB) =5.
OK!
![18](https://user-images.githubusercontent.com/52053585/206843244-42ca6179-e4ec-44af-9c5d-b52e90a4be3b.png)

11. The process Commit size on the processes tab is 11 248 K.
OK!
![19](https://user-images.githubusercontent.com/52053585/206843302-7e2ea9bf-0fd5-4ee8-afbf-b08c0caaad68.png)

12.  Process Explorer Private Bytes - 11 248 K, Virtual size 623 124 K (same picture for .NET 6.0 app). 
![20](https://user-images.githubusercontent.com/52053585/206843328-5750122b-308d-44e8-8166-c9c929f97a76.png)
OK!

End of report.

**Comment by MichalPetryka** [comment]
.Net 7 does not support Windows 7 so I don't think this will be fixed.

**Comment by ymalich** [comment]
@MichalPetryka 
> .Net 7 does not support Windows 7 so I don't think this will be fixed unless it also reproduces on supported OSes.
win7-x64 RID is still supported 
https://learn.microsoft.com/en-us/dotnet/core/rid-catalog 
![image](https://user-images.githubusercontent.com/52053585/206896070-8d878763-01b3-4ae0-8039-632a72067c16.png)

So I'd expect it should be fixed.
At least for better user-experience, to be user-friendly ;)

**Comment by KeterSCP** [comment]
@ymalich see this announcement for more information: https://github.com/dotnet/core/issues/7556

**Comment by ymalich** [comment]
@KeterSCP I know.
But I think it's better to fix the bug than not to fix it :)

**Comment by ymalich** [comment]
VS 2022 still support Windows 7 Target OS too.
![image](https://user-images.githubusercontent.com/52053585/206896582-8db9d8b2-5536-4e00-b0cf-e06c020b6d2d.png)

**Comment by KeterSCP** [comment]
@ymalich please, pay attention to this comment https://github.com/dotnet/core/issues/7556#issuecomment-1166416580

Especially part which states:
_"Won't fix" all Windows 7 specific bugs_

**Comment by teo-tsirpanis** [comment]
This is from a Hello World app in a Windows 11 machine, without disabling GC Regions or W^X. The first size column is private working set and the second is commit size.

![image](https://user-images.githubusercontent.com/12659251/207073772-783d7c50-70b3-4cd5-841c-ac3c669d0214.png)

> win7-x64 RID is still supported

@ymalich the presence of an RID does not imply that it the OS is supported. It used to be supported but not anymore and we can't remove the RID.

> But I think it's better to fix the bug than not to fix it :)

Fixing a bug is not free. Engineering effort will have to be given, and depending on the fix additional complexity might be introduced in the codebase. Not to mention that regression testing for this bug would require Windows 7 CI machines.

Given that so far the bug reproduces only on an unsupported operating system, I am going to close this issue.

**Comment by RedIODev** [comment]
I have the same problem. I run an old laptop that can't run windows 10 but windows 7 works fine. I am developing an application for this laptop specifically. The application completely halts the pc for 10 seconds before starting and afterwards hogging memory. It is unacceptable that this major issue is marked as won't fix. Not even as a low priority bug.

**Comment by ymalich** [comment]
@RedIODev 
I'm very disappointed too, because I still support all my apps for Win7 and will support them further.
There is only one way to make .NET 7 apps to work in Win7 - is to set system environment variables 
set DOTNET_GCName=clrgc.dll
set DOTNET_EnableWriteXorExecute=0
![dotnet7env](https://user-images.githubusercontent.com/52053585/210611889-34b0548a-7bff-4e06-a2f9-21778046ac66.png)
You can also use a .cmd file to launch the app under Windows 7.

@echo off

set DOTNET_GCName=clrgc.dll
set DOTNET_EnableWriteXorExecute=0

start "theApp" TheAppNet7.x64.exe

**Comment by teo-tsirpanis** [comment]
@ymalich if you want to still support Windows 7 (which you really shouldn't), your best option is to stick to .NET 6. Its support will end in November 2024 which gives ample time to migrate to a supported OS.

**Comment by RedIODev** [comment]
@ymalich that totally fixed the problem. Thank you so much.

**Comment by RedIODev** [comment]
If it is not the intention to fix the problem in the cli why not auto apply the settings the environment variables @ymalich mentioned change by default if os version 7 is detected. This would only require a minor code change but would make all dotnet 7 applications at least usable in windows 7. For my case since I only have the 1 computer running windows 7 the fix is simple (add the variables) but for programs with larger user groups requiring every windows 7 user to apply this hotfix is impractical.

**Comment by ymalich** [comment]
@teo-tsirpanis 
> which you really shouldn't
Not every PC/laptop is able  to migrate to Windows 10 and yet less PCs/laptops able to run Windows 11.
A lot of users still use Win 7 and they are happy with it. 
I read 15% Firefox users are on Windows 7 for example.
I found really bad idea to throw away technically good working hardware to the dump only because OS is not supported anymore due to political/marketing decision. 
I personally still have "old school" laptops which have not supported OSes installed and they still work well but can't run Windows 10.  

But I understand this topic/ticket is the wrong place to discuss political and marketing decisions, therefore I stop talking here..

**Comment by z4v4l** [comment]
this same behavior appears on .Net6.0 on Windows 7 x64, but not with an app, with the dotnet process instead. it hogs memory on a 8GB machine for several seconds even if run just dotnet -? setting the mentioned variables `COMPlus_GCName=clrgc.dll` and  `DOTNET_EnableWriteXorExecute=0` fixes the issue. https://github.com/dotnet/msbuild/issues/8404
